stdlib/logging/__init__.pyi

```diff
@@ -754,28 +755,26 @@ lastResort: Optional[StreamHandler]
 class StreamHandler(Handler):
     stream: SupportsWrite[str]  # undocumented
     terminator: str
-    def __init__(self, stream: Optional[SupportsWrite[str]] = ...) -> None: ...
+    def __init__(self, stream: SupportsWrite[str] | None = ...) -> None: ...
     if sys.version_info >= (3, 7):
-        def setStream(self, stream: SupportsWrite[str]) -> Optional[SupportsWrite[str]]: ...
+        def setStream(self, stream: SupportsWrite[str]) -> SupportsWrite[str] | None: ...
 
 class FileHandler(StreamHandler):
+    stream: TextIOWrapper  # undocumented
     baseFilename: str  # undocumented
     mode: str  # undocumented
-    encoding: Optional[str]  # undocumented
+    encoding: str | None  # undocumented
     delay: bool  # undocumented
     if sys.version_info >= (3, 9):
-        errors: Optional[str]  # undocumented
+        errors: str | None  # undocumented
         def __init__(
-            self,
-            filename: StrPath,
-            mode: str = ...,
-            encoding: Optional[str] = ...,
-            delay: bool = ...,
-            errors: Optional[str] = ...,
+            self, filename: StrPath, mode: str = ..., encoding: str | None = ..., delay: bool = ..., errors: str | None = ...
         ) -> None: ...
     else:
-        def __init__(self, filename: StrPath, mode: str = ..., encoding: Optional[str] = ..., delay: bool = ...) -> None: ...
-    def _open(self) -> IO[Any]: ...
+        def __init__(self, filename: StrPath, mode: str = ..., encoding: str | None = ..., delay: bool = ...) -> None: ...
+    def _open(self) -> TextIOWrapper: ...
+    if sys.version_info >= (3, 7):
+        def setStream(self, stream: TextIOWrapper) -> TextIOWrapper | None: ...  # type: ignore
```

Comment:
`FileHandler.close` still uses a getattr check before calling .close(), so arguably this class still allows other IO-like types. In practice that seems unlikely though, and we'll get some potential false positives either way. I'm fine with merging this if it makes pytest's life easier.

Reply:
This is a tradeoff. `FileHandler` opens a `TextIOWrapper` by default. While this is in theory overridable by `setStream()`, I doubt this is a common use case. (Although I have not verified this believe in any way.) I think restricting `stream` to allow only `TextIOWrapper` is the best service for users and probably correct for nearly all use cases.